apis/topology/v1beta1/topology_types.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/condition"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/affinity"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"context"
+)
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+	// +kubebuilder:validation:Optional
+	// APITopologySpreadConstraint exposes topologySpreadConstraint that are
+	// applied to the StatefulSet
+	TopologySpreadConstraint *[]corev1.TopologySpreadConstraint `json:"topologySpreadConstraint,omitempty"`
+
+	// APIAffinity exposes PodAffinity and PodAntiaffinity overrides that are applied
+	// to the StatefulSet
+	// +optional
+	APIAffinity *affinity.Overrides `json:",inline"`
```

Comment:
@stuggi `APIAffinity` (we can change the name eventually) comes from [1], where the k8s struct is wrapped to expose what we already had.
We have two choices here:
1. Expose affinity/antiaffinity from PodSpec as is
2. Refine [1] to match how we want the user to interact w/ this CR.

A Topology CR would be like the following:

```
apiVersion: topology.openstack.org/v1beta1
kind: Topology
metadata:
  name: default-sample
  namespace: openstack
spec:
  topologySpreadConstraint:
  - maxSkew: 1
    topologyKey: "topology.kubernetes.io/zone" # Spread evenly across zones
    whenUnsatisfiable: DoNotSchedule
    labelSelector:
      matchLabels:
        app: glance
 affinity:
   preferred:
      weight: 80
      topologyKey: "topology.kubernetes.io/zone"
      selectKey: "foo"
      selectValues:
        - "bar"
```
You can put only a subset of them (e.g. only define topologySpreadConstraint and skip affinity/antiaffinity).

[1] https://github.com/openstack-k8s-operators/lib-common/pull/582

Reply:
Since we don't have much size constraints here, I'm wondering if it would be a better choice to expose Affinity/AntiAffinity as provided by k8s and avoid the need of documenting these parameters. The lib-common wrapping made more sense in a scenario where we passed `affinity` as override in a form like:

```yaml
...
 glanceAPIs:
    default:
  	override:
      	    affinity:
             preferred:
          	  weight: 70
          	  selectorKey: glance
          	  selectorValues:
            	  - "glance"
            	  - "foo"
            	  - "bar"
          	  topologyKey: "topology.kubernetes.io/zone"
```

In our case we can rely on [1] (which is the same syntax exposed in StatefulSets/Deployments/etc).

[1] https://docs.openshift.com/container-platform/4.17/nodes/scheduling/nodes-scheduler-pod-affinity.html

Reply:
I cannot comment on the glance commit directly so noting it here. There is a k8s API guide section about object references that worth reading. One thing that I noted from it is that we should call the field on the glance side as topologyRef instead of just topology. 

Also looking that the examples there they always using a nested struct not just a simple string field even if the only useful field in that struct is a name field. So I'm wondering if this is a way to make the ref forward compatible when new fields are needed as adding fields to a struct is easier than replacing a string field with a struct field. 

[1]https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references

Reply:
Yes, in Glance we can follow the pattern referenced in [1] and use `topologyRef` as struct w/ all the required parameters (e.g. Namespace):

```yaml
topologyRef:
  name: foo
  namespace: foo-namespace
```
If this proposal moves forward, I will base the Glance PR on [1].

[1] https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references

Reply:
Additionally, what would be the way to define a default topology for Glance? On the Glance CRD side we cannot define a default directly now. 
* We can define a default in the glance controller code and document it at the ref field but that feels a bit hacky (even though mariadbaccount handling works similarly as if not provided the service controller creates a default one). 
* We can define a default name for the ref in Glance CRD, e.g. glance-topology and in our DTs/VAs define that topology CR with the defaults we need. (this will be a bit error prone as if somebody miss the creation of that CR Glance will refuse to deploy)
* We can add a layer of logic in the openstack-operator controller to that creates default Topology CR(s) automatically if it detects that no user defined one are passed to Glance. (this creates a bit of reversed logic that the openstack-controller defines the default of a service level field, but might be fixed to move the defaulting logic to a function in the glance api module)

Reply:
While we already have a default `Affinity` policy driven by what we defined in lib-common [1], Topology today is not generated at all.
In my mind, and because I don't need a default because [1] and `NodeSelectors` cover the default scenario, until the `openstack-operator` side of things is sorted, this would be an "advanced configuration" that you can only apply creating a separated CR and consuming it in the service operator.
Being that said:
1.  point 1 would be a no go
2. We might need something in architectures repo to match the reference architecture we would like to propose ( stretched ctlplane vs DCN vs the combination of the two)
3.  This is interesting: the simple use case would be: I want the same distribution applied to the whole ctlplane; in this case it would make sense to generate a `default` policy from openstack-operator and pass it down if there's no override at the service level.

On point 1, if I define my `topologyRef` like:

```golang
// +kubebuilder:validation:Optional
// Topology to apply the Policy defined by the associated CR referenced by name
Topology *TopologyRef `json:"topologyRef,omitempty"`
```
...
```golang
// TopologyRef -
type TopologyRef struct {
    // +kubebuilder:validation:Optional
    // Name -
    Name string `json:"name,omitempty"`
    // Namespace -
    Namespace string `json:"namespace,omitempty"`
}
```

I can still create my `Glance` CR without any Topology reference, and keep it backward compatible with my previous version (this is why it would be an advanced configuration)

[1] https://github.com/openstack-k8s-operators/lib-common/blob/main/modules/common/affinity/affinity.go

Reply:
Correct, I'm also ok to not go too far with the implementation until we have a clear view of the direction we want to take: if we define a proposal that allows the services API to grow without necessarily hitting the size limit issues we have today, and we decide to `inline` this struct into v2, we risk to double pay the price of an opinionated implementation. I think the first step in my mind is to allow services to consume a certain topology/affinity, and being able to create a CR that can be used for that purpose is the first step.